internal/painter/gl/draw.go

```diff
@@ -367,6 +429,17 @@ func (p *painter) vecRectCoords(pos fyne.Position, rect *canvas.Rectangle, frame
 	return [4]float32{x1Pos, y1Pos, x2Pos, y2Pos}, coords
 }
 
+func (p *painter) vecSquareCoords(pos fyne.Position, rect fyne.CanvasObject, frame fyne.Size) ([4]float32, []float32) {
+	xPad, yPad := float32(0), float32(0)
+	if rect.Size().Width > rect.Size().Height {
+		xPad = (rect.Size().Width - rect.Size().Height) / 2
+	} else {
+		yPad = (rect.Size().Height - rect.Size().Width) / 2
+	}
```

Comment:
It would be nice to not calculate four sizes https://github.com/fyne-io/fyne/blob/be9cd4b67990836a8d9f518b0c3b5a59b1f0349e/canvas/circle.go#L83 here each time we run. Maybe cache the size?

```suggestion
	size := rect.Size()
	if size.Width > size.Height {
		xPad = (size.Width - size.Height) / 2
	} else {
		yPad = (size.Height - size.Width) / 2
	}
```